File: python/gamex/platform.py
from __future__ import annotations
import os
from enum import Enum
from gamex.pak import PakFile
from openstk.gfx.gfx import IObjectManager, IMaterialManager, IShaderManager, ITextureManager, PlatformStats
from openstk.gfx.gfx_render import IMaterial
from openstk.gfx.gfx_texture import ITexture
from openstk.sfx.sfx import IAudioManager

# AudioBuilderBase
class AudioBuilderBase:
    def createAudio(self, path: object) -> Audio: pass
    def deleteAudio(self, audio: Audio) -> None: pass

# AudioManager
class AudioManager(IAudioManager):
    _pakFile: PakFile
    _builder: AudioBuilderBase
    _cachedAudios: dict[object, (Audio, object)] = {}
    _preloadTasks: dict[object, object] = {}

    def __init__(self, pakFile: PakFile, builder: AudioBuilderBase):
        self._pakFile = pakFile
        self._builder = builder

    def createAudio(self, key: object) -> (Audio, object):
        if path in self._cachedAudios: return self._cachedAudios[path]
        # load & cache the audio.
        tag = self._loadAudio(path)
        audio = self._builder.createAudio(tag) if tag else None
        self._cachedAudios[path] = (audio, tag)
        return (audio, tag)

    def preloadAudio(self, path: object) -> None:
        if path in self._cachedAudios: return
        # start loading the audio file asynchronously if we haven't already started.
        if not path in self._preloadTasks: self._preloadTasks[path] = self._pakFile.loadFileObject(object, path)

    def deleteAudio(self, path: object) -> None:
        if not path in self._cachedAudios: return
        self._builder.deleteAudio(self._cachedAudios[0])
        self._cachedAudios.remove(path)

    async def _loadAudio(self, path: object) -> ITexture:
        assert(not path in self._cachedAudios)
        self.preloadAudio(s)
        source = await self._preloadTasks[path]
        self._preloadTasks.remove(path)
        return source

# TextureBuilderBase
class TextureBuilderBase:
    maxTextureMaxAnisotropy: int = PlatformStats.maxTextureMaxAnisotropy
    defaultTexture: Texture
    def createTexture(self, reuse: Texture, source: ITexture, level: range = None) -> Texture: pass
    def createSolidTexture(self, width: int, height: int, rgba: list[float]) -> Texture: pass
    def createNormalMap(self, texture: Texture, strength: float) -> Texture: pass
    def deleteTexture(self, texture: Texture) -> None: pass

# TextureManager
class TextureManager(ITextureManager):
    _pakFile: PakFile
    _builder: TextureBuilderBase
    _cachedTextures: dict[object, (Texture, object)] = {}
    _preloadTasks: dict[object, ITexture] = {}

    def __init__(self, pakFile: PakFile, builder: TextureBuilderBase):
        self._pakFile = pakFile
        self._builder = builder

    def createSolidTexture(self, width: int, height: int, rgba: list[float] = None) -> Texture: return self._builder.createSolidTexture(width, height, rgba)

    def createNormalMap(self, source: Texture, strength: float) -> Texture: return self._builder.createNormalMap(source, strength)

    @property
    def defaultTexture(self) -> Texture: return self._builder.defaultTexture

    def createTexture(self, path: object, level: range = None) -> (Texture, object):
        if path in self._cachedTextures: return self._cachedTextures[path]
        # load & cache the texture.
        tag = path if isinstance(path, ITexture) else self._loadTexture(path)
        texture = self._builder.createTexture(None, tag, level) if tag else self._builder.defaultTexture
        self._cachedTextures[path] = (texture, tag)
        return (texture, tag)

    def reloadTexture(self, path: object, level: range = None) -> (Texture, object):
        if path not in self._cachedTextures: return (None, None)
        c = self._cachedTextures[path]
        self._builder.createTexture(c[0], c[1], level)
        return c

    def preloadTexture(self, path: object) -> None:
        if path in self._cachedTextures: return
        # start loading the texture file asynchronously if we haven't already started.
        if not path in self._preloadTasks: self._preloadTasks[path] = self._pakFile.loadFileObject(object, path)

    def deleteTexture(self, path: object) -> None:
        if not path in self._cachedTextures: return
        self._builder.deleteTexture(self._cachedTextures[0])
        self._cachedTextures.remove(path)

    async def _loadTexture(self, path: object) -> ITexture:
        assert(not path in self._cachedTextures)
        self.preloadTexture(s)
        source = await self._preloadTasks[path]
        self._preloadTasks.remove(path)
        return source

# ShaderBuilderBase
class ShaderBuilderBase:
    def createShader(self, path: object, args: dict[str, bool]) -> Shader: pass

# ShaderManager
class ShaderManager(IShaderManager):
    emptyArgs: dict[str, bool] = {}
    _pakFile: PakFile
    _builder: ShaderBuilderBase

    def __init__(self, pakFile: PakFile, builder: ShaderBuilderBase):
        self._pakFile = pakFile
        self._builder = builder
    
    def createShader(self, path: object, args: dict[str, bool] = None) -> (Shader, object):
        return (self._builder.createShader(path, args or self.emptyArgs), None)

# ObjectBuilderBase
class ObjectBuilderBase:
    def ensurePrefab(self) -> None: pass
    def createNewObject(self, prefab: object) -> object: pass
    def createObject(self, source: object, materialManager: IMaterialManager) -> object: pass

# ObjectManager
class ObjectManager(IObjectManager):
    _pakFile: PakFile
    _materialManager: IMaterialManager
    _builder: ObjectBuilderBase
    _cachedObjects: dict[str, object] = {}
    _preloadTasks: dict[str, object] = {}

    def __init__(self, pakFile: PakFile, materialManager: IMaterialManager, builder: ObjectBuilderBase):
        self._pakFile = pakFile
        self._materialManager = materialManager
        self._builder = builder

    def createNewObject(self, path: object) -> (object, object):
        tag = None
        self._builder.ensurePrefab()
        # load & cache the prefab.
        if not path in self._cachedObjects: prefab = self._cachedObjects[path] = (self._loadObject(path), tag)
        else: prefab = self._cachedObjects[path]
        # instantiate the prefab.
        return (self._builder.createNewObject(prefab[0]), prefab[1])
 
    def preloadObject(self, path: object) -> None:
        if path in self._cachedPrefabs: return
        # start loading the object asynchronously if we haven't already started.
        if not path in self._preloadTasks: self._preloadTasks[path] = self._pakFile.loadFileObject(object, path)

    async def _loadObject(self, path: object) -> object:
        assert(not path in self._cachedPrefabs)
        self.preloadObject(path)
        source = await self._preloadTasks[path]
        self._preloadTasks.remove(path)
        return self._builder.buildObject(source, self._materialManager)

# MaterialBuilderBase
class MaterialBuilderBase:
    textureManager : ITextureManager
    normalGeneratorIntensity: float = 0.75
    defaultMaterial: Material

    def __init__(self, textureManager: ITextureManager): self.TextureManager = textureManager
    def createMaterial(self, path: object) -> Material: pass

# MaterialManager
class MaterialManager(IMaterialManager):
    _pakFile: PakFile
    _builder: MaterialBuilderBase
    _cachedMaterials: dict[object, (Material, object)] = {}
    _preloadTasks: dict[object, IMaterial] = {}

    textureManager: ITextureManager

    def __init__(self, pakFile: PakFile, textureManager: ITextureManager, builder: MaterialBuilderBase):
        self._pakFile = pakFile
        self._textureManager = textureManager
        self._builder = builder

    def createMaterial(self, path: object) -> (Material, object):
        if path in self._cachedMaterials: return self._cachedMaterials[path]
        # load & cache the material.
        source = path if isinstance(path, IMaterial) else self._loadMaterial(path)
        material = self._builder.createMaterial(source) if source else self._builder.defaultMaterial
        tag = None #source.data if source else None
        self._cachedMaterials[path] = (material, tag)
        return (material, tag)

    def preloadMaterial(self, path: object) -> None:
        if path in self._cachedMaterials: return
        # start loading the material file asynchronously if we haven't already started.
        if not path in self._preloadTasks: self._preloadTasks[path] = self._pakFile.loadFileObject(IMaterial, path)

    async def _loadMaterial(self, path: object) -> IMaterial:
        assert(not path in self._cachedMaterials)
        self.preloadMaterial(path)
        source = await self.preloadTasks[path]
        self.preloadTasks.remove(path)
        return source

# typedefs
# class Type(Enum): pass
# class OS(Enum): pass

# Platform
class Platform:
    class Type(Enum):
        Unknown = 0
        OpenGL = 1
        Unity = 2
        Vulken = 3
        Test = 4
        Other = 5

    class OS(Enum):
        Windows = 1
        OSX = 1
        Linux = 1
        Android = 1

    platformType: Type = None
    platformTag: str = None
    platformOS: OS = OS.Windows
    gfxFactory: callable = None
    sfxFactory: callable = None
    startups: list[object] = []
    inTestHost: bool = False
    logFunc: callable = None

    class Stats:
        maxTextureMaxAnisotropy: int

    # startup
    @staticmethod
    def startup() -> None:
        if Platform.inTestHost and len(Platform.startups) == 0: Platform.startups.append(TestPlatform.startup)
        for startup in Platform.startups:
            if startup(): return
        Platform.platformType = Platform.Type.Unknown
        Platform.gfxFactory = lambda source: None
        Platform.sfxFactory = lambda source: None
        Platform.logFunc = lambda a: print(a)

# TestGfx
class TestGfx:
    def __init__(self, source):
        self._source = source

# TestGfx
class TestSfx:
    def __init__(self, source):
        self._source = source

# TestPlatform
class TestPlatform:
    @staticmethod
    def startup() -> bool:
        Platform.platformType = Platform.Type.Test
        Platform.gfxFactory = lambda source: TestGfx(source)
        Platform.sfxFactory = lambda source: TestSfx(source)
        Platform.logFunc = lambda a: print(a)
        return True

File: dotnet/Base/GameX/Platform.cs
﻿using OpenStack.Gfx;
using OpenStack.Gfx.Textures;
using OpenStack.Sfx;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using System.Threading.Tasks;
using static OpenStack.Debug;
using Stopwatch = System.Diagnostics.Stopwatch;

namespace GameX.Platforms
{
    #region AudioBuilderBase

    public abstract class AudioBuilderBase<Audio>
    {
        public abstract Audio CreateAudio(object path);
        public abstract void DeleteAudio(Audio audio);
    }

    #endregion

    #region AudioManager

    public class AudioManager<Audio>(PakFile pakFile, AudioBuilderBase<Audio> builder) : IAudioManager<Audio>
    {
        readonly PakFile PakFile = pakFile;
        readonly AudioBuilderBase<Audio> Builder = builder;
        readonly Dictionary<object, (Audio aud, object tag)> CachedAudios = [];
        readonly Dictionary<object, Task<object>> PreloadTasks = [];

        public (Audio aud, object tag) CreateAudio(object path)
        {
            if (CachedAudios.TryGetValue(path, out var c)) return c;
            // load & cache the audio.
            var tag = LoadAudio(path).Result;
            var audio = tag != null ? Builder.CreateAudio(tag) : default;
            CachedAudios[path] = (audio, tag);
            return (audio, tag);
        }

        public void PreloadAudio(object path)
        {
            if (CachedAudios.ContainsKey(path)) return;
            // start loading the texture file asynchronously if we haven't already started.
            if (!PreloadTasks.ContainsKey(path)) PreloadTasks[path] = PakFile.LoadFileObject<object>(path);
        }

        public void DeleteAudio(object path)
        {
            if (!CachedAudios.TryGetValue(path, out var c)) return;
            Builder.DeleteAudio(c.aud);
            CachedAudios.Remove(path);
        }

        async Task<object> LoadAudio(object path)
        {
            Assert(!CachedAudios.ContainsKey(path));
            PreloadAudio(path);
            var source = await PreloadTasks[path];
            PreloadTasks.Remove(path);
            return source;
        }
    }

    #endregion

    #region TextureBuilderBase

    public abstract class TextureBuilderBase<Texture>
    {
        public static int MaxTextureMaxAnisotropy
        {
            get => PlatformStats.MaxTextureMaxAnisotropy;
            set => PlatformStats.MaxTextureMaxAnisotropy = value;
        }

        public abstract Texture DefaultTexture { get; }
        public abstract Texture CreateTexture(Texture reuse, ITexture source, Range? level = null);
        public abstract Texture CreateSolidTexture(int width, int height, float[] rgba);
        public abstract Texture CreateNormalMap(Texture texture, float strength);
        public abstract void DeleteTexture(Texture texture);
    }

    #endregion

    #region TextureManager

    public class TextureManager<Texture>(PakFile pakFile, TextureBuilderBase<Texture> builder) : ITextureManager<Texture>
    {
        readonly PakFile PakFile = pakFile;
        readonly TextureBuilderBase<Texture> Builder = builder;
        readonly Dictionary<object, (Texture tex, object tag)> CachedTextures = [];
        readonly Dictionary<object, Task<ITexture>> PreloadTasks = [];

        public Texture CreateSolidTexture(int width, int height, params float[] rgba) => Builder.CreateSolidTexture(width, height, rgba);

        public Texture CreateNormalMap(Texture source, float strength) => Builder.CreateNormalMap(source, strength);

        public Texture DefaultTexture => Builder.DefaultTexture;

        public (Texture tex, object tag) CreateTexture(object path, Range? level = null)
        {
            if (CachedTextures.TryGetValue(path, out var c)) return c;
            // load & cache the texture.
            var tag = path is ITexture z ? z : LoadTexture(path).Result;
            var texture = tag != null ? Builder.CreateTexture(default, tag, level) : Builder.DefaultTexture;
            CachedTextures[path] = (texture, tag);
            return (texture, tag);
        }

        public (Texture tex, object tag) ReloadTexture(object path, Range? level = null)
        {
            if (!CachedTextures.TryGetValue(path, out var c)) return (default, default);
            Builder.CreateTexture(c.tex, (ITexture)c.tag, level);
            return c;
        }

        public void PreloadTexture(object path)
        {
            if (CachedTextures.ContainsKey(path)) return;
            // start loading the texture file asynchronously if we haven't already started.
            if (!PreloadTasks.ContainsKey(path)) PreloadTasks[path] = PakFile.LoadFileObject<ITexture>(path);
        }

        public void DeleteTexture(object path)
        {
            if (!CachedTextures.TryGetValue(path, out var c)) return;
            Builder.DeleteTexture(c.tex);
            CachedTextures.Remove(path);
        }

        async Task<ITexture> LoadTexture(object path)
        {
            Assert(!CachedTextures.ContainsKey(path));
            PreloadTexture(path);
            var source = await PreloadTasks[path];
            PreloadTasks.Remove(path);
            return source;
        }
    }

    #endregion

    #region ShaderBuilderBase

    public abstract class ShaderBuilderBase<Shader>
    {
        public abstract Shader CreateShader(object path, IDictionary<string, bool> args = null);
    }

    #endregion

    #region ShaderManager

    public class ShaderManager<Shader>(PakFile pakFile, ShaderBuilderBase<Shader> builder) : IShaderManager<Shader>
    {
        static readonly Dictionary<string, bool> EmptyArgs = [];
        readonly PakFile PakFile = pakFile;
        readonly ShaderBuilderBase<Shader> Builder = builder;

        public (Shader sha, object tag) CreateShader(object path, IDictionary<string, bool> args = null)
            => (Builder.CreateShader(path, args ?? EmptyArgs), null);
    }

    #endregion

    #region ObjectBuilderBase

    public abstract class ObjectBuilderBase<Object, Material, Texture>
    {
        public abstract void EnsurePrefab();
        public abstract Object CreateNewObject(Object prefab);
        public abstract Object CreateObject(object source, IMaterialManager<Material, Texture> materialManager);
    }

    #endregion

    #region ObjectManager

    public class ObjectManager<Object, Material, Texture>(PakFile pakFile, IMaterialManager<Material, Texture> materialManager, ObjectBuilderBase<Object, Material, Texture> builder) : IObjectManager<Object, Material, Texture>
    {
        readonly PakFile PakFile = pakFile;
        readonly IMaterialManager<Material, Texture> MaterialManager = materialManager;
        readonly ObjectBuilderBase<Object, Material, Texture> Builder = builder;
        readonly Dictionary<object, (Object obj, object tag)> CachedObjects = new Dictionary<object, (Object obj, object tag)>();
        readonly Dictionary<object, Task<object>> PreloadTasks = new Dictionary<object, Task<object>>();

        public (Object obj, object tag) CreateObject(object path)
        {
            Builder.EnsurePrefab();
            // load & cache the prefab.
            if (!CachedObjects.TryGetValue(path, out var prefab)) prefab = CachedObjects[path] = LoadObject(path).Result;
            return (Builder.CreateNewObject(prefab.obj), prefab.tag);
        }

        public void PreloadObject(object path)
        {
            if (CachedObjects.ContainsKey(path)) return;
            // start loading the object asynchronously if we haven't already started.
            if (!PreloadTasks.ContainsKey(path)) PreloadTasks[path] = PakFile.LoadFileObject<object>(path);
        }

        async Task<(Object obj, object tag)> LoadObject(object path)
        {
            Assert(!CachedObjects.ContainsKey(path));
            PreloadObject(path);
            var source = await PreloadTasks[path];
            PreloadTasks.Remove(path);
            return (Builder.CreateObject(source, MaterialManager), source);
        }
    }

    #endregion

    #region MaterialBuilderBase

    public abstract class MaterialBuilderBase<Material, Texture>(ITextureManager<Texture> textureManager)
    {
        protected ITextureManager<Texture> TextureManager = textureManager;
        public float? NormalGeneratorIntensity = 0.75f;
        public abstract Material DefaultMaterial { get; }

        public abstract Material CreateMaterial(object path);
    }

    #endregion

    #region MaterialManager

    /// <summary>
    /// Manages loading and instantiation of materials.
    /// </summary>
    public class MaterialManager<Material, Texture>(PakFile pakFile, ITextureManager<Texture> textureManager, MaterialBuilderBase<Material, Texture> builder) : IMaterialManager<Material, Texture>
    {
        readonly PakFile PakFile = pakFile;
        readonly MaterialBuilderBase<Material, Texture> Builder = builder;
        readonly Dictionary<object, (Material material, object tag)> CachedMaterials = [];
        readonly Dictionary<object, Task<IMaterial>> PreloadTasks = [];
        public ITextureManager<Texture> TextureManager { get; } = textureManager;

        public (Material mat, object tag) CreateMaterial(object path)
        {
            if (CachedMaterials.TryGetValue(path, out var c)) return c;
            // load & cache the material.
            var source = path is IMaterial z ? z : LoadMaterial(path).Result;
            var material = source != null ? Builder.CreateMaterial(source) : Builder.DefaultMaterial;
            object tag = null; // source?.Data;
            CachedMaterials[path] = (material, tag);
            return (material, tag);
        }

        public void PreloadMaterial(object path)
        {
            if (CachedMaterials.ContainsKey(path)) return;
            // start loading the material file asynchronously if we haven't already started.
            if (!PreloadTasks.ContainsKey(path)) PreloadTasks[path] = PakFile.LoadFileObject<IMaterial>(path);
        }

        async Task<IMaterial> LoadMaterial(object path)
        {
            Assert(!CachedMaterials.ContainsKey(path));
            PreloadMaterial(path);
            var source = await PreloadTasks[path];
            PreloadTasks.Remove(path);
            return source;
        }
    }

    #endregion

    #region Platform

    /// <summary>
    /// Platform
    /// </summary>
    public static class Platform
    {
        /// <summary>
        /// The platform type.
        /// </summary>
        public enum Type { Unknown, OpenGL, Unity, Unreal, Vulken, StereoKit, Test, Other }

        /// <summary>
        /// The platform OS.
        /// </summary>
        public enum OS { Windows, OSX, Linux, Android }

        /// <summary>
        /// Gets or sets the platform.
        /// </summary>
        public static Type PlatformType;

        /// <summary>
        /// Gets or sets the platform tag.
        /// </summary>
        public static string PlatformTag;

        /// <summary>
        /// Gets the platform os.
        /// </summary>
        public static readonly OS PlatformOS = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? OS.Windows
            : RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ? OS.OSX
            : RuntimeInformation.OSDescription.StartsWith("android-") ? OS.Android
            : RuntimeInformation.IsOSPlatform(OSPlatform.Linux) ? OS.Linux
            : throw new ArgumentOutOfRangeException(nameof(RuntimeInformation.IsOSPlatform), RuntimeInformation.OSDescription);

        /// <summary>
        /// Gets or sets the platforms gfx factory.
        /// </summary>
        public static Func<PakFile, IOpenGfx> GfxFactory;

        /// <summary>
        /// Gets or sets the platforms sfx factory.
        /// </summary>
        public static Func<PakFile, IOpenSfx> SfxFactory;

        /// <summary>
        /// Gets the platform startups.
        /// </summary>
        public static readonly List<Func<bool>> Startups = new List<Func<bool>>();

        /// <summary>
        /// Determines if in a test host.
        /// </summary>
        public static bool InTestHost => AppDomain.CurrentDomain.GetAssemblies().Any(x => x.FullName.StartsWith("testhost,"));

        /// <summary>
        /// The platform stats.
        /// </summary>
        public class Stats
        {
            static readonly bool HighRes = Stopwatch.IsHighResolution;
            static readonly double HighFrequency = 1000.0 / Stopwatch.Frequency;
            static readonly double LowFrequency = 1000.0 / TimeSpan.TicksPerSecond;
            static readonly bool UseHrt = false;
            public static bool UsingHighResolutionTiming => UseHrt && HighRes && !Unix;
            public static long TickCount => (long)Ticks;
            public static double Ticks => UseHrt && HighRes && !Unix ? Stopwatch.GetTimestamp() * HighFrequency : DateTime.UtcNow.Ticks * LowFrequency;
            public static readonly bool Is64Bit = Environment.Is64BitProcess;
            public static bool MultiProcessor { get; private set; }
            public static int ProcessorCount { get; private set; }
            public static bool Unix { get; private set; }
            public static bool VR { get; private set; }
        }

        public static void Startup()
        {
            if (InTestHost && Startups.Count == 0) Startups.Add(TestPlatform.Startup);
            foreach (var startup in Startups) if (startup()) return;
            PlatformType = Type.Unknown;
            GfxFactory = source => null; // throw new Exception("No GfxFactory");
            SfxFactory = source => null; // throw new Exception("No SfxFactory");
            AssertFunc = x => System.Diagnostics.Debug.Assert(x);
            LogFunc = a => System.Diagnostics.Debug.Print(a);
            LogFormatFunc = (a, b) => System.Diagnostics.Debug.Print(a, b);
        }
    }

    #endregion

    #region TestGfx

    public interface ITestGfx : IOpenGfx { }

    public class TestGfx(PakFile source) : ITestGfx
    {
        readonly PakFile _source = source;
        public object Source => _source;
        public Task<T> LoadFileObject<T>(object path) => throw new NotSupportedException();
        public void PreloadTexture(object path) => throw new NotSupportedException();
        public void PreloadObject(object path) => throw new NotSupportedException();
    }

    #endregion

    #region TestSfx

    public interface ITestSfx : IOpenSfx { }

    public class TestSfx(PakFile source) : ITestSfx
    {
        readonly PakFile _source = source;
        public object Source => _source;
    }

    #endregion

    #region TestPlatform

    public static class TestPlatform
    {
        public static bool Startup()
        {
            try
            {
                Platform.PlatformType = Platform.Type.Test;
                Platform.GfxFactory = source => new TestGfx(source);
                Platform.SfxFactory = source => new TestSfx(source);
                AssertFunc = x => System.Diagnostics.Debug.Assert(x);
                LogFunc = a => System.Diagnostics.Debug.Print(a);
                LogFormatFunc = (a, b) => System.Diagnostics.Debug.Print(a, b);
                return true;
            }
            catch { return false; }
        }
    }

    #endregion
}